src/app/fdctl/topology.c

```diff
@@ -0,0 +1,731 @@
+    ADD( link_wksp );
+  }
+
+  if( FD_LIKELY( tile->out_link_id_primary!=ULONG_MAX ) ) {
+    fd_topo_wksp_t * link_wksp = &topo->workspaces[ topo->links[ tile->out_link_id_primary ].wksp_id ];
+    ADD( link_wksp );
+  }
+
+  fd_topo_wksp_t * tile_wksp = &topo->workspaces[ tile->wksp_id ];
+  ADD( tile_wksp );
+
+  if( FD_UNLIKELY( tile->kind == FD_TOPO_TILE_KIND_SHRED ) ) {
+    /* Shred tile locks 5 scratch pages for storing private key material. */
+    tile_mem += 5UL * FD_SHMEM_NORMAL_PAGE_SZ;
+  }
+
+  return tile_mem;
+}
+
+ulong
+fd_topo_mlock_max_tile( fd_topo_t * topo ) {
+  fd_topo_fill( topo, FD_TOPO_FILL_MODE_FOOTPRINT );
+
+  ulong result = 0UL;
+  for( ulong i=0; i<topo->tile_cnt; i++ ) {
+    fd_topo_tile_t * tile = &topo->tiles[ i ];
+    result = fd_ulong_max( result, fd_topo_mlock_max_tile1( topo, tile ) );
+  }
+
+  return result;
+}
+
+ulong
+fd_topo_mlock_all( fd_topo_t * topo ) {
+  fd_topo_fill( topo, FD_TOPO_FILL_MODE_FOOTPRINT );
+
+  ulong result = 0UL;
+  for( ulong i=0; i<topo->wksp_cnt; i++ ) {
+    result += topo->workspaces[ i ].page_cnt * topo->workspaces[ i ].page_sz;
+  }
+  return result;
+}
+
+void
+fd_topo_validate( fd_topo_t * topo ) {
+  /* Tiles have valid wksp_ids */
+  for( ulong i=0; i<topo->tile_cnt; i++ ) {
+    if( FD_UNLIKELY( topo->tiles[ i ].wksp_id >= topo->wksp_cnt ) )
+      FD_LOG_ERR(( "invalid workspace id %lu", topo->tiles[ i ].wksp_id ));
+  }
+
+  /* Links have valid wksp_ids */
+  for( ulong i=0; i<topo->link_cnt; i++ ) {
+    if( FD_UNLIKELY( topo->links[ i ].wksp_id >= topo->wksp_cnt ) )
+      FD_LOG_ERR(( "invalid workspace id %lu", topo->links[ i ].wksp_id ));
+  }
+
+  /* Tiles of the same kind share the same wksp id */
+  for( ulong i=0; i<topo->tile_cnt; i++ ) {
+    for( ulong j=0; j<topo->tile_cnt; j++ ) {
+      if( FD_UNLIKELY( i==j ) ) continue;
+      if( topo->tiles[ i ].kind == topo->tiles[ j ].kind ) {
+        if( FD_UNLIKELY( topo->tiles[ i ].wksp_id != topo->tiles[ j ].wksp_id ) )
+          FD_LOG_ERR(( "tiles %lu and %lu of kind %lu have different wksp ids", i, j, topo->tiles[ i ].kind ));
+      }
+    }
+  }
+
+  /* Links of the same kind share the same wksp id */
+  for( ulong i=0; i<topo->link_cnt; i++ ) {
+    for( ulong j=0; j<topo->link_cnt; j++ ) {
+      if( FD_UNLIKELY( i==j ) ) continue;
+      if( topo->links[ i ].kind == topo->links[ j ].kind ) {
+        if( FD_UNLIKELY( topo->links[ i ].wksp_id != topo->links[ j ].wksp_id ) )
+          FD_LOG_ERR(( "links %lu and %lu of kind %lu have different wksp ids", i, j, topo->links[ i ].kind ));
+      }
+    }
+  }
+
+  /* Tiles of different kind have different workspace ids */
+  for( ulong i=0; i<topo->tile_cnt; i++ ) {
+    for( ulong j=0; j<topo->tile_cnt; j++ ) {
+      if( FD_UNLIKELY( i==j ) ) continue;
+      if( topo->tiles[ i ].kind != topo->tiles[ j ].kind ) {
+        if( FD_UNLIKELY( topo->tiles[ i ].wksp_id == topo->tiles[ j ].wksp_id ) )
```

Comment:
You could combine this with the previous loop easily, but maybe it's nice for error message sake
(kind==kind)^(id==id)

Reply:
Yeah I prefer having these all separate